src/services/MouseController.cxx

```diff
@@ -0,0 +1,69 @@
+#include "MouseController.hxx"
+#include "../util/LOG.hxx"
+#include "../util/iShape.hxx"
+#include "../util/Rectangle.hxx"
+
+MouseController::MouseController(GameService::ServiceTuple & context) :
+	GameService(context), m_LastHovered(nullptr)
+{
+	LOG(LOG_DEBUG) << "Created MouseController service";
+}
+
+MouseController::~MouseController()
+{
+  LOG(LOG_DEBUG) << "Destroying MouseController";
+}
+
+void MouseController::handleEvent(MousePositionEvent && event)
+{
+  Point2D point{event.xPosition, event.yPosition};
+  if(m_LastHovered)
+  {
+    if(m_LastHovered->getShape().contains(point)) return;
+    else m_LastHovered->onMouseLeave();
+  }
+  SpatialBlock block{event};
+  if(m_SpatialMap.count(block) > 0)
+  {
+    for(auto handler : m_SpatialMap.at(block))
+      if(handler->getShape().contains(point))
+      {
+        handler->onMouseHover();
+        m_LastHovered = handler;
+        return;
+      }
+  }
+  m_LastHovered = nullptr;
+}
+
+void MouseController::addHandler(iMouseHandler * handler)
+{
+  Rectangle bounds = handler->getShape().getBounds();
+  LOG(LOG_DEBUG) << "Adding a mouse handler for region " << bounds;
+  auto [x1, y1] = bounds.p1();
+  auto [x2, y2] = bounds.p2();
+  constexpr int block = SpatialBlock::BlockSize::value;
+  for(int i = x1 / block; i <= x2 / block; ++i)
+    for(int j = y1 / block; j <= y2 / block; ++j)
+    {
+      LOG(LOG_DEBUG) << "Adding it to region (" << i << ", " << j << ")";
+      m_SpatialMap[SpatialBlock{i, j}].push_back(handler);
+    }
+}
+
+void MouseController::handleEvent(ClickEvent && event)
+{
+  if(m_LastHovered) m_LastHovered->onMouseClick(std::move(event));
+}
+
+void MouseController::handleEvent(ScrollEvent && event)
+{
+  if(m_LastHovered) m_LastHovered->onScroll(std::move(event));
```

Comment:
Would it be possible that after scrolling positions of elements are changed? So we are not hovering above the same element. Would call to the MousePositionEvent handler has sense?

Reply:
Yes. You are right. This is a big flaw with my code right now. I think that the solution is that elements contained in iLayouts must be relative to the iLayout element itself.

Reply:
Probably, what the solution will have to be is that each iLayout will own a MouseController.

Reply:
And MouseController will no longer be a service. Instead, the Window's MouseController will be used instead just like I did to MouseState

Reply:
I think this is a big enough problem, that it should be included in this PR

Reply:
Ok so I spent some time thinking about this. See discord messages for what I will do. But tl;dr when we are scrolling, we are scrolling the element itself, not the whole window. So if the element itself has sub-components, its iMouseHandler will have to handle the proper dispatching.

Reply:
This work will be done in Pt II